quickwit/quickwit-control-plane/src/indexing_plan.rs

```diff
@@ -0,0 +1,657 @@
+// Copyright (C) 2022 Quickwit, Inc.
+//
+// Quickwit is offered under the AGPL v3.0 and as commercial software.
+// For commercial licensing, contact us at hello@quickwit.io.
+//
+// AGPL:
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as
+// published by the Free Software Foundation, either version 3 of the
+// License, or (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program. If not, see <http://www.gnu.org/licenses/>.
+
+use std::cmp::Ordering;
+use std::collections::HashMap;
+
+use itertools::Itertools;
+use quickwit_cluster::ClusterMember;
+use quickwit_config::{SourceConfig, CLI_INGEST_SOURCE_ID, INGEST_API_SOURCE_ID};
+use quickwit_proto::indexing_api::IndexingTask;
+use serde::Serialize;
+
+/// A [`PhysicalIndexingPlan`] defines the list of indexing tasks
+/// each indexer, identified by its node ID, should run.
+/// TODO(fmassot): a metastore version number will be attached to the plan
+/// to identify if the plan is up to date with the metastore.
+#[derive(Debug, PartialEq, Clone, Serialize)]
+pub struct PhysicalIndexingPlan {
+    indexing_tasks_per_node_id: HashMap<String, Vec<IndexingTask>>,
```

Comment:
This function exposes a lot of public methods with quirky contract.

I'd feel safer if there was a builder and the actual plan was just a read only object with a restricted number of methods.

Reply:
I removed the hidden contract stuff and foot guns (asserts)...
I did not go for the builder as it seems safer now without asserts and expects.